### ISSUE TITLE
App Localization

### ISSUE THREAD

**Issue by a-polivanchuk** [strongbox-password-safe/Strongbox#137]
Please add support for translation in order to have Strongbox available in different languages.

**Comment by mmcguill** [comment]
Yes, definitely. Just waiting on the UI to settle over the next few weeks, it's a bit of a moving target right now but this is planned. Thanks

**Comment by a-polivanchuk** [comment]
Hi! I see there is a file Localizable.strings available. I'd like to start translating it. What do you think?
Which localization platform are you considering to use?

**Comment by mmcguill** [comment]
Hi! I’m using GitLocalize, well it’s currently under test with Russian! But it looks usable! What language did you want to work on?

**Comment by a-polivanchuk** [comment]
I'd like to translate into Ukrainian! Could you please add it for the test too?

**Comment by mmcguill** [comment]
Hi @Slummi - you can find them in the Babel project here: https://github.com/strongbox-password-safe/babel

Please let me know if you want to work on the German project, I'll add you to GitLocalize.

@a-polivanchuk - I'll add you shortly to the Ukrainian localization - We just need to make sure everything goes well for Russian and write a guide as it's a bit involved, and a lot of work!

**Comment by mmcguill** [comment]
If anyone else is out there and wants to contribute translations for their language please get in touch and I'll try to set you up :)

**Comment by Slummi** [comment]
Hi @mmcguill Yes, I‘d like to work on the German localization, but I haven‘t worked with GitLocalize before. So I have to look how it works. Are there separate localizations for the different platforms (MacOS, iOS and now iPadOS) or is it one project? I‘m currently using iOS only, so iOS would be the only OS where I‘m able to have a look into the app to check any contexts, if this is necessary.

**Comment by mmcguill** [comment]
Hi @Slummi - Only the iOS app is localized, so that's all you'd have to look at. I've added you as a moderator for German now. Have a look around, it's a fairly straightforward system:

https://gitlocalize.com/repo/3101

I'm working on putting together a guide/FAQ at the moment, but feel free to submit Pull Requests for anything you think is best

**Comment by Slummi** [comment]
Hi @mmcguill - Had a quick look on the GitLocalize project. Can you tell me, why there are some strange translations (see red boxes in the screenshot).
Is there something wrong with the assignment between GitLocalize and the related string files in the repository? Looks like the object identifiers...

![2019_09_25_20_07_12_GitLocalize_Continuous_Localization_for_GitHub_Projects](https://user-images.githubusercontent.com/42265381/65627933-aff87c80-dfd0-11e9-96cf-527b2c937e1b.png)

**Comment by mmcguill** [comment]
Yes, GitLocalize isn't brilliant, and it seems isn't optimised for XCode...

You can ignore these, and I'll mention them in the guide.

**Comment by Slummi** [comment]
Ok, so these are no real strings to translate and I must only translate strings where the translation is empty or English or there‘s a better translation for an existing one. Right?

**Comment by mmcguill** [comment]
For the moment, that's the conclusion yes... Still working it out, but in general what you say sounds right.

**Comment by a-polivanchuk** [comment]
Hello!
Just submitted first PRs via GitLocalize.

I'm curious, have you considered any other options for localization, for instance, Crowdin? It has GitHub integration with auto submitting new translations, branches support, translation memory and it's fully compatible with iOS l10n files.

**Comment by mmcguill** [comment]
Hi @a-polivanchuk, thanks! 

Yes, it was one of the first I looked at, unfortunately it's quite expensive it seems. The project has around 6000 strings, which would mean $150-$300 a month. 

I don't think Gitlocalize is great, but it's not so bad. The initial translation process is painful but after that it's pretty straightforward to update or add a couple of strings.

Could you drop me a mail at support@strongboxsafe.com to discuss the process?

**Comment by mmcguill** [comment]
I've updated the README in Babel with some hints, tips, and advice we've gained from the Russian translation if you'd like to take a look.

**Comment by a-polivanchuk** [comment]
@Slummi I've already created German locale on Crowdin and imported existing translations.
Please join your locale by using this invitation link:
https://crowdin.com/project/strongbox-password-safe/invite?d=c695v4r4o466f56343i5m4s4a3m4q4e3

We together with @mmcguill decided to set up the localization process on Crowdin and see how it works.
It would be good if you continue your work there and give your feedback :)

Thanks!

**Comment by mmcguill** [comment]
NB: @Slummi - things still work with GitLocalize also, but it seems like CrowdIn is a better interface entirely. So feel free to choose the way you prefer to contribute. I'll message our Russian translator separately to let him know he can also switch

**Comment by Slummi** [comment]
OK, I signed up for Crowdin and will test it.

**Comment by Slummi** [comment]
Ok, I translated "Upgrade.strings" with Crowdin for testing. But I can't find any option to upload the changes or make a pull request. Is it not necessary to make a PR with Crowdin?

**Comment by a-polivanchuk** [comment]
It’s not necessary to do anything but translating and approving :) Everything will be pulled and pushed automatically.

**Comment by Slummi** [comment]
Ok and what happens next after approving the changes? Can‘t see any PRs or updates to the repository.

**Comment by a-polivanchuk** [comment]
@Slummi The new commits will be pushed automatically during each sync, which is now set to 24 hours.
The new PR will be created automatically too.

**Comment by a-polivanchuk** [comment]
BTW, Crowdin has a mobile version which works perfectly on mobile browsers.
So it's possible to translate even on the go.

**Comment by mmcguill** [comment]
Thanks @a-polivanchuk - I think we're good to close this off now. We've got the process for Localization all setup in CrowdIn. Cheers everyone.

**Comment by Slummi** [comment]
Even if this one has already been closed out, just a short note.

It seems that some strings are still missing on Crowdin, and that some strings have been translated, but they stay in English in the app.

The statistics in the database management for example are in English again, but they have already been translated a long time ago.

And the strings on the webpage of the "Transfer over LAN" feature are completly missing on Crowdin.
![2019-10-24 14_44_32-Window](https://user-images.githubusercontent.com/42265381/67486833-0b1f9c80-f66d-11e9-8bd7-d000f0d1ff68.png)

**Comment by a-polivanchuk** [comment]
I've searched the strings from your screenshot on Crowdin and there were no results. The TM doesn't contain those strings too. So it looks like they have never been on Crowdin.
I guess we need to search for those original strings on GitHub files and check the changesets to figure out the issue...

**Comment by mmcguill** [comment]
ok, thanks guys, I'll look into this today and get back to you

**Comment by mmcguill** [comment]
So, the strings from the above are going to be a little awkward because it's html/javascript rather than objective-C and so they need to be injected into the html dynamically.

I can make the long description string available in the next release for localization easily because it's already injected and I don't need to do much work

**Comment by mmcguill** [comment]
I do, and it's good or bad news depending on how you look at it. It appears to be down to a bug in XCode 11.1, so another Apple bug... It seems like there is some serious Quality Control problems in Apple with their software in general at the moment, sad to say. Details here:

https://developer.apple.com/documentation/xcode_release_notes/xcode_11_1_release_notes/#3386781
https://stackoverflow.com/questions/57905965/xcode-11-ios-13-localization-issue

So, the good news is it's not my fault and will be fixed hopefully in 11.2... I'm not sure when that releases though.

I think the only thing we can do is wait for that release, rebuild and re-assess then. :(

**Comment by Slummi** [comment]
Yes, Apple has a major software quality problem I think. iOS 13 is complete garbage if you ask me. It has so many bugs...
So let's wait and hope that it will become better. In the meanwhile I found additional strings, which are not translated in the app. :(

**Comment by xatage** [comment]
Yes, Apple's latest iOS and macOS are a real disaster if you look at all the bugs and troubles the users are faced with. 
I was just yesterday contacting Mark about this issue. It was really a lot of effort to figure out what was the cause of missing translation at some places in the App. Now we can only hope, Apple brings out the Xcode update quickly to fix this translation issue.
